### PR TITLE
Ownership Voucher certificate chain verification

### PR DESCRIPTION
**Pull request by Darshini-Parikh93**
- Performing ownership voucher certificate chain verification for
public key type ECDSA.

Signed-off-by: Darshini Parikh <darshini.k.parikh@intel.com>

### CODE REVIEW

File: protocol/src/main/java/org/fido/iot/protocol/To0ServerService.java [hunk 1]
```diff
@@ -47,8 +47,13 @@ protected void doOwnerSign(Composite request, Composite reply) {
     byte[] to0dNonce3 = to0d.getAsBytes(Const.TO0D_NONCE3);
     to0d.verifyMaxKey(Const.TO0D_NONCE3);
 
-    //todo:verify voucher
-    //cryptoService.verifyVoucher(voucher);
+    // todo:verify voucher
```
Comment (johneasterday) [inline protocol/src/main/java/org/fido/iot/protocol/To0ServerService.java:50]: Do we still need this todo comment?

Comment (Darshini-Parikh93) [reply protocol/src/main/java/org/fido/iot/protocol/To0ServerService.java]: Removed `todo` from the comment.

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java [hunk 2]
```diff
@@ -579,6 +581,113 @@ public void verify(Composite chain) {
...
+    }
+  }
+
+  /**
+   * Verifies certificate public key size.
+   *
+   * @param pubKeySize Public key size
+   */
+  private void verifyKeySize(int pubKeySize) throws InvalidOwnershipVoucherException {
+    if (pubKeySize != Const.BIT_LEN_256 && pubKeySize != Const.BIT_LEN_384) {
```
Comment (larkinscarvalho) [inline protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:615]: should be or condition?
`pubKeySize != Const.BIT_LEN_256 || pubKeySize != Const.BIT_LEN_384`?

Comment (Darshini-Parikh93) [reply protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:615]: The proposed `||` condition in the if statement would throw exception even if one of the condition is not satisfied. Let's say key size is 256 bits in that case the second condition will fail and throw exception and when key size is 384 bits, the first condition will fail and throw exception. However the purpose behind this check is to throw exception when key size is neither 256 bits or 384 bits, so `&&` condition is appropriate.

Comment (larkinscarvalho) [reply protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:615]: Thanks for the explanation!
I misread not equal to condition.

Comment (Darshini-Parikh93) [reply protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:615]: No problem!!

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java [hunk 3]
```diff
@@ -579,6 +581,113 @@ public void verify(Composite chain) {
...
+      }
+    }
+  }
+
+  /**
+   * Verifies certificate chain.
+   *
+   * @param certChain ownership voucher device certificate chain
+   */
+  public void verifyCertChain(Composite certChain) {
```
Comment (prateekcr) [inline protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:658]: Functionally, it verifies the leaf cert, and not the entire cert chain, unless the method 'verify()' is called within this method. The method could be renamed in the future.

**Review by prateekcr — APPROVED** [review]

**Review by trbehera — APPROVED** [review]

**Review by johneasterday — APPROVED** [review]

**Pull request merged**